### PR TITLE
feat: data plane view support

### PR DESCRIPTION
**Pull request by wenytang-ms**
[feature design](https://microsoft.sharepoint.com/:fl:/s/a17c993b-3d61-4d51-8538-13bf2d5982de/EWdmmQ6z6ERFu8dCr3BDAykBkgt6ijTmdt3h45kaooTYaw?e=phJw2j&nav=cz0lMkZzaXRlcyUyRmExN2M5OTNiLTNkNjEtNGQ1MS04NTM4LTEzYmYyZDU5ODJkZSZkPWIlMjE1S2F3RWZPV0NrLVRQeXV6c0cxQU9WTUtrNWp3MFU1Q2g0b3MzLW1jS1dOQlRMSkJxb19EUXFpcW1TT196VzB3JmY9MDFIRU9ISFhESE0yTVE1TTdJSVJDM1hSMkNWNVlFR0FaSiZjPSUyRiZhPUxvb3BBcHAmcD0lNDBmbHVpZHglMkZsb29wLXBhZ2UtY29udGFpbmVyJng9JTdCJTIydyUyMiUzQSUyMlQwUlRVSHh0YVdOeWIzTnZablF1YzJoaGNtVndiMmx1ZEM1amIyMThZaUUxUzJGM1JXWlBWME5yTFZSUWVYVjZjMGN4UVU5V1RVdHJOV3AzTUZVMVEyZzBiM016TFcxalMxZE9RbFJNU2tKeGIxOUVVWEZwY1cxVFQxOTZWekIzZkRBeFNFVlBTRWhZUjFSR1dWQTFVelZVUkVJMVF6TkxSVlJHV1ZSRFVUZERWVTglM0QlMjIlMkMlMjJpJTIyJTNBJTIyNGE4MmE5YmEtNmU5OC00ODU4LTkxY2EtMTI1ZjZlMmZmMWY3JTIyJTdE)

### CODE REVIEW

File: src/commands/registerApiSubCommands/registerStepByStep.ts [hunk 1]
```diff
@@ -16,6 +17,9 @@ export async function registerStepByStep(context: IActionContext, node?: ApisTre
         const apiCenterNode = await ext.treeDataProvider.showTreeItemPicker<ApiCenterTreeItem>(ApiCenterTreeItem.contextValue, context);
         node = apiCenterNode.apisTreeItem;
     }
+    if (node.apiCenter instanceof ApiCenterVersionDefinitionManagement) {
```
Comment (formulahendry) [inline src/commands/registerApiSubCommands/registerStepByStep.ts:20]: return when `ApiCenterVersionDefinitionManagement`?

File: src/extension.ts [hunk 2]
```diff
@@ -59,6 +65,9 @@ export async function activate(context: vscode.ExtensionContext) {
 
     const azureAccountTreeItem = createAzureAccountTreeItem(sessionProvider);
     context.subscriptions.push(azureAccountTreeItem);
+    ext.treeItem = azureAccountTreeItem;
+    // var a = ext.treeItem.subscription;
```
Comment (formulahendry) [inline src/extension.ts:69]: removed unused code

File: src/extension.ts [hunk 3]
```diff
@@ -33,14 +37,16 @@ import { openRule } from './commands/rules/openRule';
...
 import { testInPostman } from './commands/testInPostman';
-import { ErrorProperties, TelemetryProperties } from './common/telemetryEvent';
 import { doubleClickDebounceDelay, selectedNodeKey } from './constants';
 import { ext } from './extensionVariables';
 import { ApiVersionDefinitionTreeItem } from './tree/ApiVersionDefinitionTreeItem';
 import { createAzureAccountTreeItem } from "./tree/AzureAccountTreeItem";
+import { createAzureDataAccountTreeItem } from './tree/DataPlaneAccount';
 import { OpenApiEditor } from './tree/Editors/openApi/OpenApiEditor';
-
+// Copilot Chat
```
Comment (formulahendry) [inline src/extension.ts:48]: removed unused code

File: src/extension.ts [hunk 4]
```diff
@@ -68,6 +77,17 @@ export async function activate(context: vscode.ExtensionContext) {
     const treeView = vscode.window.createTreeView("apiCenterTreeView", { treeDataProvider });
     context.subscriptions.push(treeView);
 
+    AzureDataSessionProviderHelper.activateAzureSessionProvider(context);
```
Comment (formulahendry) [inline src/extension.ts:80]: Better to wrap those initialazation logic for data plane into a function

File: src/commands/addDataPlaneApis.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    }
+    const clientid = await vscode.window.showInputBox({ title: UiStrings.AddDataPlaneClientId, ignoreFocusOut: true });
+    if (!clientid) {
+        return;
+    }
+    const tenantid = await vscode.window.showInputBox({ title: UiStrings.AddDataPlaneTenantId, ignoreFocusOut: true });
+    if (!tenantid) {
+        return;
+    }
+    // return await getSessionToken(clientid, tenantid);
```
Comment (formulahendry) [inline src/commands/addDataPlaneApis.ts:21]: remove unused code

File: src/commands/addDataPlaneApis.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    setAccountToExt(endpointUrl, clientid, tenantid);
+    ext.workspaceItem.refresh(context);
+}
+export function setAccountToExt(domain: string, clientId: string, tenantId: string) {
+    function pushIfNotExist(array: DataPlaneAccount[], element: DataPlaneAccount) {
+        if (!array.some(item => item.domain === element.domain)) {
+            array.push(element);
+        }
+    }
+    pushIfNotExist(ext.dataPlaneAccounts, { domain: domain, tenantId: tenantId, clientId: clientId });
```
Comment (formulahendry) [inline src/commands/addDataPlaneApis.ts:31]: Better to show warning message when domain is same?

Comment (wenytang-ms) [reply src/commands/addDataPlaneApis.ts:33]: good suggestion

File: src/extension.ts (hunk 4 above)
Comment (wenytang-ms) [reply src/extension.ts]: wrapper both for data plane and control plane

**Review by sthirthala — COMMENTED** [review]
I have reviewed the package.nls and uistrings

This PR has other things which I have not reviewed.

File: src/commands/registerApiSubCommands/registerStepByStep.ts [hunk 7]
```diff
@@ -17,7 +17,7 @@ export async function registerStepByStep(context: IActionContext, node?: ApisTre
         const apiCenterNode = await ext.treeDataProvider.showTreeItemPicker<ApiCenterTreeItem>(ApiCenterTreeItem.contextValue, context);
```
Comment (formulahendry) [inline src/commands/registerApiSubCommands/registerStepByStep.ts:17]: I rememeber `showTreeItemPicker` could pass the param with a start tree item, so with that, end user would not be able to select a DataPlane tree item.

File: src/utils/generalUtils.ts [hunk 8]
```diff
@@ -54,4 +54,15 @@ export namespace GeneralUtils {
...
         return path.join(os.homedir(), apiCenterFolder);
     }
+
+    export async function fetchDataFromLink(link: string): Promise<string> {
+        try {
+            const res = await fetch(link);
+            const rawData = await res.json();
+            return JSON.stringify(rawData);
+        }
+        catch (err) {
```
Comment (formulahendry) [inline src/utils/generalUtils.ts:64]: try-catch-throw seems redundant?

File: src/utils/generalUtils.ts [hunk 9]
```diff
@@ -1,8 +1,8 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT license.
+import fetch from 'node-fetch';
```
Comment (formulahendry) [inline src/utils/generalUtils.ts:3]: we already have `axios` as dependency. better to reuse existing dependency or Node.js built-in lib?

File: src/extensionVariables.ts [hunk 10]
```diff
@@ -17,4 +18,8 @@ export namespace ext {
     export let outputChannel: IAzExtOutputChannel;
     export let openApiEditor: OpenApiEditor;
     export let selectedApiVersionDefinitionTreeItem: ApiVersionDefinitionTreeItem;
+
+    export let dataPlaneAccounts: DataPlaneAccount[];
+    export let workspaceProvider: AzExtTreeDataProvider;
+    export let workspaceItem: AzExtParentTreeItem & { dispose(): unknown; };
```
Comment (formulahendry) [inline src/extensionVariables.ts:24]: Better naming for `workspace` related to Data Plane. `workspace` is a different concept in API Center.

File: src/commands/exportApi.ts [hunk 11]
```diff
@@ -1,32 +1,25 @@
...
 import { ApiVersionDefinitionTreeItem } from "../tree/ApiVersionDefinitionTreeItem";
 import { createTemporaryFolder } from "../utils/fsUtil";
+import { GeneralUtils } from "../utils/generalUtils";
 export namespace ExportAPI {
     export async function exportApi(
         context: IActionContext,
         node?: ApiVersionDefinitionTreeItem): Promise<void> {
         if (!node) {
-            node = await ext.treeDataProvider.showTreeItemPicker<ApiVersionDefinitionTreeItem>(new RegExp(`${ApiVersionDefinitionTreeItem.contextValue}*`), context);
+            node = await ext.treeDataProvider.showTreeItemPicker<ApiVersionDefinitionTreeItem>(new RegExp(`${ApiCenterVersionDefinitionManagement.contextValue}*`), context);
```
Comment (formulahendry) [inline src/commands/exportApi.ts:19]: So from Command Palette, user could only select a tree item in Management Plane View?
When user only have Data Plane view, the Command Palette would be useless to user?

File: src/commands/registerApiSubCommands/registerStepByStep.ts (hunk 7 above)
Comment (wenytang-ms) [reply src/commands/registerApiSubCommands/registerStepByStep.ts:17]: the tree view from ext.treeDataProvider, this treeDataProvider is for control plane, data plane is named as workspaceProvider

Comment (formulahendry) [reply src/commands/registerApiSubCommands/registerStepByStep.ts:17]: So, L20-L22 is not needed?

Comment (wenytang-ms) [reply src/commands/registerApiSubCommands/registerStepByStep.ts:17]: ok

File: src/utils/generalUtils.ts (hunk 8 above)
Comment (wenytang-ms) [reply src/utils/generalUtils.ts:64]: you say redundant with `registerWithTelemetry` catch logic?

File: src/commands/exportApi.ts (hunk 11 above)
Comment (wenytang-ms) [reply src/commands/exportApi.ts]: updated

File: package.json [hunk 12]
```diff
@@ -441,6 +441,7 @@
     ],
     "configuration": [
       {
+        "title": "Azure-Api-Center",
```
Comment (formulahendry) [inline package.json:444]: Better to be `Azure API Center`？ As it will be shown to end user in VS Code settings.

File: src/uiStrings.ts [hunk 13]
```diff
@@ -125,4 +125,6 @@ export class UiStrings {
     static readonly RequestFailedWithStatusCode = vscode.l10n.t("Request failed with status code: {0}");
     static readonly DownloadDefinitionFileWithErrorMsg = vscode.l10n.t("Download API Center Definition File error: {0}");
     static readonly DatplaneAlreadyAdded = vscode.l10n.t("This Data Plane Runtime URL already added to Data View.");
+    static readonly SelectItemFromTreeView = vscode.l10n.t("Select from which tree view");
+    static readonly GetTreeView = vscode.l10n.t("Please Sign In to Azure to get management view or connect to data plane APIs");
```
Comment (formulahendry) [inline src/uiStrings.ts]: Please ask Sreek to help review all newly added strings.

`Sign In` -> `Sign in`

Better naming for `management view`

File: src/constants.ts [hunk 14]
```diff
@@ -34,6 +34,11 @@ export const RegisterApiOptions = {
     cicd: UiStrings.RegisterApiOptionCicd,
 };
 
+export const TreeViewType = {
+    controlPlaneView: "API Center Management View",
```
Comment (formulahendry) [inline src/constants.ts:38]: Put into UiStrings.ts

File: src/commands/exportApi.ts [hunk 15]
```diff
@@ -6,17 +6,25 @@ import * as fs from "fs-extra";
...
+import { UiStrings } from "../uiStrings";
 import { createTemporaryFolder } from "../utils/fsUtil";
 import { GeneralUtils } from "../utils/generalUtils";
 export namespace ExportAPI {
     export async function exportApi(
         context: IActionContext,
         node?: ApiVersionDefinitionTreeItem): Promise<void> {
         if (!node) {
-            node = await ext.treeDataProvider.showTreeItemPicker<ApiVersionDefinitionTreeItem>(new RegExp(`${ApiCenterVersionDefinitionManagement.contextValue}*`), context);
+            let res = await getDefinitionTreeNode(context);
```
Comment (formulahendry) [inline src/commands/exportApi.ts:23]: Do we need `res`? Just `node = await getDefinitionTreeNode(context);` is OK?

File: src/commands/exportApi.ts [hunk 16]
```diff
@@ -49,4 +57,33 @@ export namespace ExportAPI {
         await vscode.workspace.fs.writeFile(vscode.Uri.file(localFilePath), Buffer.from(fileContent));
         await vscode.window.showTextDocument(document);
     }
+
+    async function getDefinitionTreeNode(context: IActionContext): Promise<ApiVersionDefinitionTreeItem | null> {
```
Comment (formulahendry) [inline src/commands/exportApi.ts:61]: In our scope for the first release of data plane, do we have other commands that will trigger from Command Palette?

Even though ExportAPI is the only one for first release, we'd better to put this shared method in another file that all command could reuse.

**Review by formulahendry — CHANGES_REQUESTED** [review]
Overall looks good now. Just some small comments.

File: src/commands/exportApi.ts (hunk 15 above)
Comment (wenytang-ms) [reply src/commands/exportApi.ts:23]: no, the getDefinitionTreeNode return type is node|null,  so can not assign null to node.

File: package.json (hunk 12 above)
Comment (wenytang-ms) [reply package.json]: sure

File: src/uiStrings.ts (hunk 13 above)
Comment (wenytang-ms) [reply src/uiStrings.ts]: update this sentence to "Please connect to Azure API Center Service first"

File: src/commands/exportApi.ts (hunk 16 above)
Comment (wenytang-ms) [reply src/commands/exportApi.ts]: updated.

File: src/uiStrings.ts [hunk 17]
```diff
@@ -127,4 +127,6 @@ export class UiStrings {
     static readonly DatplaneAlreadyAdded = vscode.l10n.t("This Data Plane Runtime URL already added to Data View.");
     static readonly SelectItemFromTreeView = vscode.l10n.t("Select from which tree view");
     static readonly GetTreeView = vscode.l10n.t("Please connect to Azure API Center Service first.");
+    static readonly APIControlPlaneView = vscode.l10n.t("API Center Management View");
```
Comment (formulahendry) [inline src/uiStrings.ts:130]: "API Center Management Plane View" to algin with "API Center Data Plane View"?

Comment (formulahendry) [reply src/uiStrings.ts:130]: Is it shown in picker of Command Palette? Do we really need to add 'View' in string?

File: src/commands/exportApi.ts (hunk 15 above)
Comment (formulahendry) [reply src/commands/exportApi.ts]: node is `typeX | undefiend`, while getDefinitionTreeNode is `typeX | null`.

There is another related comment: getDefinitionTreeNode is better to return `typeX | undefiend`. I rarely see a fucntion return `typeX | null`, but more are `typeX | undefiend`.

With above impreovement, we could also remove `res`.

Comment (wenytang-ms) [reply src/commands/exportApi.ts]: okay

File: src/uiStrings.ts (hunk 17 above)
Comment (wenytang-ms) [reply src/uiStrings.ts]: change to API Center Management Plane and API Center Data Plane

**Review by formulahendry — APPROVED** [review]

**Review by sthirthala — APPROVED** [review]

File: package.nls.json
Comment (sthirthala) [inline package.nls.json:1]: Instead of saying "Remove API Center" should it be "Disconnect from API Center"

**Review by sthirthala — COMMENTED** [review]
Added comment - we should likely change Remove API Center to be Disconnect from API Center

File: package.nls.json
Comment (wenytang-ms) [reply package.nls.json:1]: thank you! the string updated.

**Pull request merged**